checkov/gitlab_ci/common/resource_id_utils.py

```diff
@@ -3,16 +3,54 @@
 
 from checkov.common.util.consts import START_LINE, END_LINE
 
+IMAGE_BLOCK_NAMES = ('image', 'services')
+
 
 def generate_resource_key_recursive(conf: dict[str, Any] | list[str] | str, key: str, start_line: int,
                                     end_line: int) -> str:
+    return _generate_resource_key_recursive(conf, key, start_line, end_line, set(), 0)
+
+
+def _generate_resource_key_recursive(conf: dict[str, Any] | list[str] | str, key: str, start_line: int,
+                                     end_line: int, scanned_image_blocks: set[str], depth: int) -> str:
     if not isinstance(conf, dict):
         return key
 
     for k, value in conf.items():
+        if k in IMAGE_BLOCK_NAMES:
+            scanned_image_blocks.add(k)
+
         if isinstance(value, dict) and value[START_LINE] <= start_line <= end_line <= value[END_LINE]:
             next_key = f'{key}.{k}' if key else k
-            return generate_resource_key_recursive(value, next_key, start_line, end_line)
+            return _generate_resource_key_recursive(value, next_key, start_line, end_line, scanned_image_blocks,
+                                                    depth + 1)
+
+        if isinstance(value, list) and isinstance(value[0], dict):
+            next_key = f'{key}.{k}' if key else k
+
+            for idx, value_dict in enumerate(value):
+                if value_dict[START_LINE] <= start_line <= end_line <= value_dict[END_LINE]:
+                    next_key += f'.{idx + 1}'
+                    break  # There can be only one match in terms of line range
+
+            return _generate_resource_key_recursive(value, next_key, start_line, end_line, scanned_image_blocks,
+                                                    depth + 1)
+
+        if any(block_name in conf.keys()
```

Comment:
Can this be extracted out of the encapsulating for loop?
Seems we're doing the same for every conf iteration

Reply:
we need to do it at every iteration because we don't know in what depth these block names will appear